test/e2e/es/certs_test.go

```diff
@@ -32,24 +32,161 @@ import (
+						_ = k.Client.Delete(context.Background(), &toDelete)
+					},
+				},
+			}
+		},
 	}
 
+	// The initial Cluster builder should result in a healthy cluster as verified by the standard check steps
+	// Now modify cluster to use custom HTTP CA but simulate a user error by populating the secret with
+	// garbage and verify this is bubbled up through an event
+	b := initialCluster.
+		WithCustomHTTPCerts(caSecretName).
+		WithMutatedFrom(&initialCluster)
+
+	withBogusCert := test.WrappedBuilder{
+		BuildingThis: b,
+
+		// we don't want to run a regular mutation as the invalid certificates will not lead to a successful upgrade
+		OverrideMutationSteps: func(k *test.K8sClient) test.StepList {
+			steps := test.StepList{
+				{
+					Name: "Create an invalid CA secret",
+					Test: test.Eventually(func() error {
+						bogusSecret := mkCertSecret([]byte("garbage"), []byte("more garbage"))
+						_, err := reconciler.ReconcileSecret(k.Client, bogusSecret, nil)
+						return err
+					}),
+				},
+			}
+			steps = append(steps, b.UpgradeTestSteps(k)...)
+			steps = append(steps,
+				test.Step{
+					Name: "Invalid CA secret should create events",
+					Test: test.Eventually(func() error {
+						eventList, err := k.GetEvents(test.EventListOptions(esNamespace, initialCluster.Elasticsearch.Name)...)
+						if err != nil {
+							return err
+						}
+						for _, evt := range eventList {
+							if evt.Type == corev1.EventTypeWarning &&
+								evt.Reason == events.EventReconciliationError &&
+								evt.InvolvedObject.Namespace == b.Elasticsearch.Namespace &&
+								evt.InvolvedObject.Name == b.Elasticsearch.Name &&
+								strings.Contains(evt.Message, "can't parse") {
+								return nil
+							}
+						}
+						return fmt.Errorf("expected validation event but could not observe it")
+					}),
+				},
+			)
+			return steps
+		},
+	}
+
+	// A NOOP mutation but set up the CA secret correctly now (using the existing CA generation code in the operator)
+	withCustomCert := test.WrappedBuilder{
+		BuildingThis: initialCluster.
+			WithCustomHTTPCerts(caSecretName).
+			WithMutatedFrom(&initialCluster),
+
+		PreMutationSteps: func(k *test.K8sClient) test.StepList {
+			return test.StepList{
+				{
+					Name: "Update custom CA secret",
+					Test: test.Eventually(func() error {
+						var err error
+						customCA, err = certificates.NewSelfSignedCA(certificates.CABuilderOptions{
+							Subject: pkix.Name{
+								CommonName:         "eck-e2e-test-custom-ca",
+								OrganizationalUnit: []string{"eck-e2e"},
+							},
+						})
+						if err != nil {
+							return err
+						}
+
+						caSecret := mkCertSecret(
+							certificates.EncodePEMCert(customCA.Cert.Raw),
+							certificates.EncodePEMPrivateKey(*customCA.PrivateKey),
+						)
+						_, err = reconciler.ReconcileSecret(k.Client, caSecret, nil)
+						return err
+					}),
+				},
+			}
+		},
+		PostMutationSteps: func(k *test.K8sClient) test.StepList {
```

Comment:
Any reason for having this check as a `PostMutationStep` ? I think we should expect the CA to be available right after the update, something along those lines:

```golang
PreMutationSteps: func(k *test.K8sClient) test.StepList {
    return test.StepList{
        {
            Name: "Update custom CA secret",
            Test: test.Eventually(func() error {
                var err error
                customCA, err = certificates.NewSelfSignedCA(certificates.CABuilderOptions{
                    Subject: pkix.Name{
                        CommonName:         "eck-e2e-test-custom-ca",
                        OrganizationalUnit: []string{"eck-e2e"},
                    },
                })
                if err != nil {
                    return err
                }

                caSecret := mkCertSecret(
                    certificates.EncodePEMCert(customCA.Cert.Raw),
                    certificates.EncodePEMPrivateKey(*customCA.PrivateKey),
                )
                _, err = reconciler.ReconcileSecret(k.Client, caSecret, nil)
                return err
            }),
        },
        {
            // Updated CA should be exposed almost immediately 
            Name: "Verify that the custom CA is in use",
            Test: test.Eventually(func() error {
                return elasticsearch.CheckHTTPConnectivityWithCA(initialCluster.Elasticsearch, k, []*x509.Certificate{customCA.Cert})
            }),
        },
    }
},
```

Reply:
Good point!

Reply:
I guess the only reason would be that it allows you to switch the sequence of the builders without breaking the test.